src/pages/Dashboard/Modals/SupplyWithdraw/SupplyWithdrawForm.tsx

```diff
@@ -0,0 +1,202 @@
+/** @jsxImportSource @emotion/react */
+import React, { useState, useEffect } from 'react';
+import BigNumber from 'bignumber.js';
+import { FormikProps } from 'formik';
+import { Typography } from '@mui/material';
+import { FormValues } from 'containers/AmountForm/validationSchema';
+import { AmountForm, IAmountFormProps } from 'containers/AmountForm';
+import {
+  TokenTextField,
+  Delimiter,
+  LabeledInlineContent,
+  ILabeledInlineContentProps,
+  PrimaryButton,
+  ProgressBarHorizontal,
+  Tooltip,
+  ValueUpdate,
+} from 'components';
+import { useTranslation } from 'translation';
+import { Asset, TokenId } from 'types';
+import {
+  getBigNumber,
+  convertCoinsToWei,
+  convertWeiToCoins,
+  formatCentsToReadableValue,
+  format,
+  currencyFormatter,
+} from 'utilities/common';
+import { useStyles } from '../styles';
+
+interface ISupplyWithdrawFormProps {
+  asset: Asset;
+  tokenInfo: ILabeledInlineContentProps[];
+  userTotalBorrowBalance: BigNumber;
+  userTotalBorrowLimit: BigNumber;
+  dailyEarnings: BigNumber;
+  onSubmit: IAmountFormProps['onSubmit'];
+  enabledButtonKey: string;
+  disabledButtonKey: string;
+  maxInput: BigNumber;
+  calculateNewBalance: (amount: BigNumber) => BigNumber;
+}
+
+export const SupplyWithdrawContent: React.FC<
+  Omit<ISupplyWithdrawFormProps, 'onSubmit'> & FormikProps<FormValues>
+> = ({
+  values,
+  setFieldValue,
+  asset,
+  tokenInfo,
+  userTotalBorrowBalance,
+  userTotalBorrowLimit,
+  dailyEarnings,
+  maxInput,
+  enabledButtonKey,
+  disabledButtonKey,
+  calculateNewBalance,
+}) => {
+  const { id: assetId } = asset;
+  const { amount } = values;
+  const validAmount = amount && !amount.isZero() && !amount.isNaN();
+  const styles = useStyles();
+  const { t, Trans } = useTranslation();
+  const [newBorrowLimit, setNewBorrowLimit] = useState<BigNumber | undefined>(undefined);
+  const [newBorrowPercent, setNewBorrowPercent] = useState(
+    userTotalBorrowBalance.div(userTotalBorrowLimit).times(100),
+  );
+  useEffect(() => {
+    const tokenPrice = getBigNumber(asset?.tokenPrice);
+    const collateralFactor = getBigNumber(asset?.collateralFactor);
+
+    if (tokenPrice && validAmount) {
+      const amountInUsd = convertWeiToCoins({
+        value: amount as BigNumber,
+        tokenSymbol: asset.id,
+      })
+        .times(tokenPrice)
+        .times(collateralFactor);
+      const temp = calculateNewBalance(amountInUsd);
+      setNewBorrowLimit(BigNumber.maximum(temp, 0));
+      setNewBorrowPercent(userTotalBorrowBalance.div(temp).times(100));
+    } else if (BigNumber.isBigNumber(userTotalBorrowLimit)) {
+      setNewBorrowLimit(undefined);
+      setNewBorrowPercent(userTotalBorrowBalance.div(userTotalBorrowLimit).times(100));
+    }
+  }, [amount, asset?.id, userTotalBorrowBalance, userTotalBorrowLimit]);
```

Comment:
I can't respond to your previous comment about using `useMemo` for optimizations only. I agree with that, however my point was that this code doesn't need a state to be written (so it could also be written without `useMemo`):
```typescript
let uncappedBorrowLimit = userTotalBorrowLimit;
let borrowLimit: BigNumber | undefined;

if (tokenPrice && validAmount) {
  const tokenPrice = getBigNumber(asset?.tokenPrice);
  const collateralFactor = getBigNumber(asset?.collateralFactor);

  const amountInUsd = convertWeiToCoins({
    value: amount as BigNumber,
    tokenSymbol: asset.id,
  })
    .times(tokenPrice)
    .times(collateralFactor);

  uncappedBorrowLimit = calculateNewBalance(amountInUsd);
  borrowLimit = BigNumber.maximum(uncappedBorrowLimit, 0);
}

const borrowPercent = userTotalBorrowBalance.div(uncappedBorrowLimit).times(100).toNumber();
```

Because you're not updating `newBorrowLimit` or `newBorrowLimit` from within the component expecting it to react to the change, you don't need to store these values in states.
The `else if` statement isn't needed in your logic, you're already initializing these variables with these values.

Now using `useMemo` means the calculation of these two variables wouldn't happen on every render, thus optimizing the process a bit:

```typescript
const [newBorrowLimit, newBorrowPercent] = React.useMemo(() => {
  let uncappedBorrowLimit = userTotalBorrowLimit;
  let borrowLimit: BigNumber | undefined;

  if (tokenPrice && validAmount) {
    const tokenPrice = getBigNumber(asset?.tokenPrice);
    const collateralFactor = getBigNumber(asset?.collateralFactor);

    const amountInUsd = convertWeiToCoins({
      value: amount as BigNumber,
      tokenSymbol: asset.id,
    })
      .times(tokenPrice)
      .times(collateralFactor);

    uncappedBorrowLimit = calculateNewBalance(amountInUsd);
    borrowLimit = BigNumber.maximum(uncappedBorrowLimit, 0);
  }

  const borrowPercent = userTotalBorrowBalance.div(uncappedBorrowLimit).times(100).toNumber();

  return [borrowLimit, borrowPercent];
}, [asset?.tokenPrice, asset?.collateralFactor, userTotalBorrowBalance, userTotalBorrowLimit]);
```